### PR TITLE
Add a PolkaVM-based executor

### PR DESCRIPTION
**Pull request by koute**
This PR adds a new PolkaVM-based executor to Substrate.

- The executor can now be used to actually run a PolkaVM-based runtime, and successfully produces blocks.
- The executor is always compiled-in, but is disabled by default.
- The `SUBSTRATE_ENABLE_POLKAVM` environment variable must be set to `1` to enable the executor, in which case the node will accept both WASM and PolkaVM program blobs (otherwise it'll default to WASM-only). This is deliberately undocumented and not explicitly exposed anywhere (e.g. in the command line arguments, or in the API) to disincentivize anyone from enabling it in production. If/when we'll move this into production usage I'll remove the environment variable and do it "properly".
- I did not use our legacy runtime allocator for the PolkaVM executor, so currently every allocation inside of the runtime will leak guest memory until that particular instance is destroyed. The idea here is that I will work on the https://github.com/polkadot-fellows/RFCs/pull/4 which will remove the need for the legacy allocator under WASM, and that will also allow us to use a proper non-leaking allocator under PolkaVM.
- I also did some minor cleanups of the WASM executor and deleted some dead code.

No prdocs included since this is not intended to be an end-user feature, but an unofficial experiment, and shouldn't affect any current production user. Once this is production-ready a full Polkadot Fellowship RFC will be necessary anyway.

### CODE REVIEW

**Review by michalkucharczyk — APPROVED** [review]

File: substrate/client/executor/common/src/lib.rs [hunk 1]
```diff
@@ -25,3 +25,9 @@ pub mod error;
 pub mod runtime_blob;
 pub mod util;
 pub mod wasm_runtime;
+
+pub(crate) fn is_polkavm_enabled() -> bool {
+	std::env::var_os("SUBSTRATE_ENABLE_POLKAVM")
+		.map(|value| value == "1")
+		.unwrap_or(false)
```
Comment (michalkucharczyk) [inline substrate/client/executor/common/src/lib.rs:32]: nit: map_or?

**Review by dmitry-markin — APPROVED** [review]

**Review by davxy — APPROVED** [review]
LGTM. I suppose that at some point there will be a follow up to rename `WasmError` type, `WasmInstance` trait, ...

File: substrate/client/executor/common/src/lib.rs (hunk 1 above)
Comment (koute) [reply substrate/client/executor/common/src/lib.rs:32]: I always forget this... (:

File: substrate/client/executor/common/src/runtime_blob/runtime_blob.rs [hunk 2]
```diff
@@ -187,20 +177,57 @@ impl RuntimeBlob {
...
+		match self.0 {
+			BlobKind::WebAssembly(raw_module) => Ok(raw_module),
+			BlobKind::PolkaVM(..) => Err(WasmError::Other(
+				"expected a WebAssembly program; found a PolkaVM program blob".into(),
+			)),
+		}
+	}
+
+	/// Gets a reference to the inner PolkaVM program blob, if this is a PolkaVM program.
+	pub fn as_polkavm_blob(&self) -> Option<&polkavm::ProgramBlob> {
```
[the remark this answers is not in the document]
Comment (koute) [reply substrate/client/executor/common/src/runtime_blob/runtime_blob.rs:227]: Even though they might be named in a similar way they're meant for different things. (:

The `as_webassembly_blob`/`as_webassembly_blob_mut` are used purely internally (they're private) to make things more convenient and have the error messages is one place. The `as_polkavm_blob` is public and is used externally (well, externally-ish) by the executor crate to extract the blob (`if let Some(blob) = blob.as_polkavm_blob() {`) so it was a deliberate decision to not have this return a `Result`.

File: substrate/client/executor/polkavm/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+		state_args.sbrk(raw_data_length);
+		state_args.reset_memory(true);
+
+		match self.0.update_state(state_args) {
+			Ok(()) => {},
+			Err(polkavm::ExecutionError::Trap(trap)) => {
+				return (Err(format!("call into the runtime method '{name}' failed: failed to allocate memory for the input payload: {trap}").into()), None);
+			},
+			Err(polkavm::ExecutionError::Error(error)) => {
+				return (Err(format!("call into the runtime method '{name}' failed: failed to allocate memory for the input payload: {error}").into()), None);
```
[the remark this answers is not in the document]
Comment (koute) [reply substrate/client/executor/polkavm/src/lib.rs:71]: The main purpose of this call is to prepare memory for the subsequent execution, but you do have a point; I'll rewrite the error message.

**Pull request merged**